streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -227,6 +230,27 @@ public void initializeIfNeeded() {
         }
     }
 
+    private void initOffsetsIfNeeded(final java.util.function.Consumer<Set<TopicPartition>> offsetResetter) {
```

Comment:
I was considering to maybe merge this method into `initMetadata()` but it might convolute different code path, and we should execute this method rarely anyway so I don't think we should have concerns about calling `mainConsumer.committed` twice for rare cases.

Let me know what you think.

Reply:
Hm...I'm not necessarily that concerned about calling `mainConsumer.committed` twice in rare cases (although maybe that would not be so good, since those rare cases happen to be those in which this is probably more likely to time out, right?)
But personally, just coming into this code from the outside, it's super confusing to have two different methods for initializing the offsets. It seems more convoluted that way, to me. Also maybe I am missing some context here but why do we call `initOffsetsIfNeeded`  from `initializeIfNeeded` rather than from `completeRestoration` in the first place? We don't need to initialize main consumer offsets until it transitions to running

Reply:
Good point!